src/herder/HerderImpl.cpp

```diff
@@ -726,19 +734,44 @@ HerderImpl::sendSCPStateToPeer(uint32 ledgerSeq, Peer::pointer peer)
     ZoneScoped;
     bool log = true;
     auto maxSlots = Herder::LEDGER_VALIDITY_BRACKET;
+
+    auto sendSlot = [=](SCPEnvelope const& e, bool log) {
+        StellarMessage m;
+        m.type(SCP_MESSAGE);
+        m.envelope() = e;
+        auto mPtr = std::make_shared<StellarMessage const>(m);
+        peer->sendMessage(mPtr, log);
+        return true;
+    };
+
+    bool delayCheckpoint = false;
+    auto checkpoint = getMostRecentCheckpointSeq();
+    auto highestSlot = getSCP().getHighestKnownSlotIndex();
+
+    // If there is a gap between the latest completed checkpoint and the next
+    // saved message, we should delay sending the checkpoint ledger. Send all
+    // other messages first, then send checkpoint messages after node that is
+    // catching up knows network state
+    if (highestSlot - mApp.getConfig().MAX_SLOTS_TO_REMEMBER - 1 > checkpoint)
+    {
+        delayCheckpoint = true;
+    }
+
+    // Send MAX_SLOTS_TO_SEND slots
     getSCP().processSlotsAscendingFrom(ledgerSeq, [&](uint64 seq) {
+        if (seq == checkpoint && delayCheckpoint)
```

Comment:
I think this is left over from when I was doing some weird things with CatchupManager that I missed during the simplification pass, I'll take it out

Reply:
Sorry, now I remember why we do this. We special case message filtering if the message is a checkpoint ledger (see `src/herder/HerderImpl.cpp:628`). This checkpoint ledger is based on `trackingConsensusLedgerIndex`. However, if the checkpoint message is the first message sent, the node does not yet have the correct `trackingConsensusLedgerIndex`, so it cannot determine that the message is the special case checkpoint message that should not be filtered. I'll add a better comment explaining this, if it took me a little while to remember this it definetely needs some better documentation. This delay makes it so that the receiving node probably has the correct `trackingConsensusLedgerIndex` by the time it receives the checkpoint message

Reply:
hmm, why is this a problem though? as long as we send sequential ledgers without gaps, the delay shouldn't matter, right?

Reply:
If we send sequential ledgers, the checkpoint message always gets sent first. This gets thrown out though since the close time of the checkpoint ledger is larger than `MAXIMUM_LEDGER_CLOSETIME_DRIFT`. We need at least one message that won't get filtered out by `MAXIMUM_LEDGER_CLOSETIME_DRIFT` to give us a baseline ledger number to calculate the correct checkpoint ledger so we can special case the filtering for messages for that checkpoint ledger

Reply:
>If we send sequential ledgers, the checkpoint message always gets sent first

Just to clarify, it's not always first, right? It could be in the middle of the sequence of ledgers that we're sending

> This gets thrown out though since the close time difference of the checkpoint ledger is larger than MAXIMUM_LEDGER_CLOSETIME_DRIFT. We need at least one message that won't get filtered out by MAXIMUM_LEDGER_CLOSETIME_DRIFT to give us a baseline ledger number to calculate the correct checkpoint ledger so we can special case the filtering for messages for that checkpoint ledger. The delay gives us reasonable probability that at least one other message is processed before the checkpoint message is received

Ohh I see, it's the close time drift (at first I was only thinking about ledger numbers, where we discard ledgers with a gap). Interesting, so the closetime drift is 50 seconds, yet by default we ask other nodes to send us is the last 12 slots, or ~60 seconds of consensus. So we basically reject a few oldest slots anyway (regardless of checkpoint logic), which doesn't quite make sense. I'm inclined to update `MAXIMUM_LEDGER_CLOSETIME_DRIFT` to be expressed in terms of MAX_SLOTS_TO_REMEMBER and average ledger close time. But we don't have to do it in this PR; even then, there will still be a possibility that checkpoint ledger is rejected. So let's just keep the delay then 👍